md2conf/converter.py

```diff
@@ -142,20 +153,60 @@ def _change_ext(path: str, target_ext: str) -> str:
 class ConfluenceStorageFormatConverter(NodeVisitor):
     "Transforms a plain HTML tree into the Confluence storage format."
 
+    path: str
     base_path: str
     links: List[str]
     images: List[str]
+    page_metadata: Dict[str, ConfluencePageMetadata]
 
-    def __init__(self, base_path: str) -> None:
+    def __init__(self, path: str, page_metadata: Dict[str, ConfluencePageMetadata] = dict()) -> None:
         super().__init__()
-        self.base_path = base_path
+        self.path = path
+        self.base_path = os.path.dirname(path)
         self.links = []
         self.images = []
+        self.page_metadata = page_metadata
 
     def _transform_link(self, anchor: ET.Element) -> ET.Element:
         url = anchor.attrib["href"]
         if not is_absolute_url(url):
+            LOGGER.debug(f"found link {url} relative to {self.path}")
             self.links.append(url)
+            relative_url = urlparse(url)
+
+            # Convert the relative href to absolute based on relative url the base path value then
+            # look up the absolute path in the page metadata dictionary to discover
+            # the relative path within confluence that should be used
+            if not relative_url.path:
+                # relative url with no path are relative to the current path
+                abs_path = self.path
+            else:
+                abs_path = os.path.abspath(
+                    os.path.join(os.path.abspath(self.base_path), relative_url.path)
+                )
+
+            # lookup page metadata using the page path
+            link_metadata = self.page_metadata.get(abs_path)
+            transformed_url = None
+            if link_metadata:
+                LOGGER.debug(f"found page {abs_path} with metadata: {link_metadata}")
+                transformed_url = f"https://{link_metadata.domain}/wiki/spaces/{link_metadata.space_key}/pages/{link_metadata.page_id}"
+
+                if relative_url.fragment:
+                    # confluence url's use `+` instead of spaces for the page title part
+                    confluence_page_title = link_metadata.title.replace(" ", "+")
```

Comment:
Instead of ad-hoc string manipulation, I would use the built-in Python package `urllib.parse`. Specifically, this package has [urlencode](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode), which not only replaces spaces with `+` but also escapes other special characters. Also, I would suggest [urlunparse](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlunparse) to construct URLs from parts, not piece them together as strings, which can be error prone.